File: app/src/main/java/dev/ragnarok/fenrir/api/impl/LocalServerApi.java
package dev.ragnarok.fenrir.api.impl;

import static dev.ragnarok.fenrir.util.Objects.nonNull;

import dev.ragnarok.fenrir.api.ILocalServerServiceProvider;
import dev.ragnarok.fenrir.api.interfaces.ILocalServerApi;
import dev.ragnarok.fenrir.api.model.Items;
import dev.ragnarok.fenrir.api.model.VKApiAudio;
import dev.ragnarok.fenrir.api.model.VKApiPhoto;
import dev.ragnarok.fenrir.api.model.VKApiVideo;
import dev.ragnarok.fenrir.api.model.response.BaseResponse;
import dev.ragnarok.fenrir.util.Utils;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.exceptions.Exceptions;
import io.reactivex.rxjava3.functions.Function;

class LocalServerApi implements ILocalServerApi {

    private final ILocalServerServiceProvider service;

    LocalServerApi(ILocalServerServiceProvider service) {
        this.service = service;
    }

    static <T> Function<BaseResponse<T>, T> extractResponseWithErrorHandling() {
        return response -> {
            if (nonNull(response.error)) {
                throw Exceptions.propagate(new Exception(Utils.firstNonEmptyString(response.error.errorMsg, "Error")));
            }

            return response.response;
        };
    }

    @Override
    public Single<Items<VKApiVideo>> getVideos(Integer offset, Integer count, boolean reverse) {
        return service.provideLocalServerService()
                .flatMap(service -> service.getVideos(offset, count, reverse ? 1 : 0)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<Items<VKApiAudio>> getAudios(Integer offset, Integer count, boolean reverse) {
        return service.provideLocalServerService()
                .flatMap(service -> service.getAudios(offset, count, reverse ? 1 : 0)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<Items<VKApiPhoto>> getPhotos(Integer offset, Integer count, boolean reverse) {
        return service.provideLocalServerService()
                .flatMap(service -> service.getPhotos(offset, count, reverse ? 1 : 0)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<Items<VKApiAudio>> getDiscography(Integer offset, Integer count, boolean reverse) {
        return service.provideLocalServerService()
                .flatMap(service -> service.getDiscography(offset, count, reverse ? 1 : 0)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<Items<VKApiVideo>> searchVideos(String query, Integer offset, Integer count, boolean reverse) {
        return service.provideLocalServerService()
                .flatMap(service -> service.searchVideos(query, offset, count, reverse ? 1 : 0)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<Items<VKApiPhoto>> searchPhotos(String query, Integer offset, Integer count, boolean reverse) {
        return service.provideLocalServerService()
                .flatMap(service -> service.searchPhotos(query, offset, count, reverse ? 1 : 0)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<Items<VKApiAudio>> searchAudios(String query, Integer offset, Integer count, boolean reverse) {
        return service.provideLocalServerService()
                .flatMap(service -> service.searchAudios(query, offset, count, reverse ? 1 : 0)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<Items<VKApiAudio>> searchDiscography(String query, Integer offset, Integer count, boolean reverse) {
        return service.provideLocalServerService()
                .flatMap(service -> service.searchDiscography(query, offset, count, reverse ? 1 : 0)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<Integer> update_time(String hash) {
        return service.provideLocalServerService()
                .flatMap(service -> service.update_time(hash)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<Integer> delete_media(String hash) {
        return service.provideLocalServerService()
                .flatMap(service -> service.delete_media(hash)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<String> get_file_name(String hash) {
        return service.provideLocalServerService()
                .flatMap(service -> service.get_file_name(hash)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<Integer> update_file_name(String hash, String name) {
        return service.provideLocalServerService()
                .flatMap(service -> service.update_file_name(hash, name)
                        .map(extractResponseWithErrorHandling()));
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/api/impl/LocalServerApi.kt
package dev.ragnarok.fenrir.api.impl

import dev.ragnarok.fenrir.UserAgentTool
import dev.ragnarok.fenrir.api.ILocalServerServiceProvider
import dev.ragnarok.fenrir.api.interfaces.ILocalServerApi
import dev.ragnarok.fenrir.api.model.Items
import dev.ragnarok.fenrir.api.model.VKApiAudio
import dev.ragnarok.fenrir.api.model.VKApiPhoto
import dev.ragnarok.fenrir.api.model.VKApiVideo
import dev.ragnarok.fenrir.api.model.response.BaseResponse
import dev.ragnarok.fenrir.model.FileRemote
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.util.Utils.firstNonEmptyString
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flatMapConcat
import kotlinx.coroutines.flow.map

internal class LocalServerApi(private val service: ILocalServerServiceProvider) : ILocalServerApi {
    override fun getVideos(offset: Int?, count: Int?, reverse: Boolean): Flow<Items<VKApiVideo>> {
        return service.provideLocalServerService()
            .flatMapConcat {
                it.getVideos(offset, count, if (reverse) 1 else 0)
                    .map(extractResponseWithErrorHandling())
            }
    }

    override fun getAudios(offset: Int?, count: Int?, reverse: Boolean): Flow<Items<VKApiAudio>> {
        return service.provideLocalServerService()
            .flatMapConcat {
                it.getAudios(offset, count, if (reverse) 1 else 0)
                    .map(extractResponseWithErrorHandling())
            }
    }

    override fun getPhotos(offset: Int?, count: Int?, reverse: Boolean): Flow<Items<VKApiPhoto>> {
        return service.provideLocalServerService()
            .flatMapConcat {
                it.getPhotos(offset, count, if (reverse) 1 else 0)
                    .map(extractResponseWithErrorHandling())
            }
    }

    override fun getDiscography(
        offset: Int?,
        count: Int?,
        reverse: Boolean
    ): Flow<Items<VKApiAudio>> {
        return service.provideLocalServerService()
            .flatMapConcat {
                it.getDiscography(offset, count, if (reverse) 1 else 0)
                    .map(extractResponseWithErrorHandling())
            }
    }

    override fun searchVideos(
        query: String?,
        offset: Int?,
        count: Int?,
        reverse: Boolean
    ): Flow<Items<VKApiVideo>> {
        return service.provideLocalServerService()
            .flatMapConcat {
                it.searchVideos(query, offset, count, if (reverse) 1 else 0)
                    .map(extractResponseWithErrorHandling())
            }
    }

    override fun searchPhotos(
        query: String?,
        offset: Int?,
        count: Int?,
        reverse: Boolean
    ): Flow<Items<VKApiPhoto>> {
        return service.provideLocalServerService()
            .flatMapConcat {
                it.searchPhotos(query, offset, count, if (reverse) 1 else 0)
                    .map(extractResponseWithErrorHandling())
            }
    }

    override fun searchAudios(
        query: String?,
        offset: Int?,
        count: Int?,
        reverse: Boolean
    ): Flow<Items<VKApiAudio>> {
        return service.provideLocalServerService()
            .flatMapConcat {
                it.searchAudios(query, offset, count, if (reverse) 1 else 0)
                    .map(extractResponseWithErrorHandling())
            }
    }

    override fun searchDiscography(
        query: String?,
        offset: Int?,
        count: Int?,
        reverse: Boolean
    ): Flow<Items<VKApiAudio>> {
        return service.provideLocalServerService()
            .flatMapConcat {
                it.searchDiscography(query, offset, count, if (reverse) 1 else 0)
                    .map(extractResponseWithErrorHandling())
            }
    }

    override fun update_time(hash: String?): Flow<Int> {
        return service.provideLocalServerService()
            .flatMapConcat {
                it.update_time(hash)
                    .map(extractResponseWithErrorHandling())
            }
    }

    override fun delete_media(hash: String?): Flow<Int> {
        return service.provideLocalServerService()
            .flatMapConcat {
                it.delete_media(hash)
                    .map(extractResponseWithErrorHandling())
            }
    }

    override fun get_file_name(hash: String?): Flow<String> {
        return service.provideLocalServerService()
            .flatMapConcat {
                it.get_file_name(hash)
                    .map(extractResponseWithErrorHandling())
            }
    }

    override fun update_file_name(hash: String?, name: String?): Flow<Int> {
        return service.provideLocalServerService()
            .flatMapConcat {
                it.update_file_name(hash, name)
                    .map(extractResponseWithErrorHandling())
            }
    }

    override fun powerOffPC(type: String): Flow<Int> {
        return service.provideLocalServerService()
            .flatMapConcat {
                it.powerOffPC(type)
                    .map(extractResponseWithErrorHandling())
            }
    }

    override fun fsGet(dir: String?): Flow<Items<FileRemote>> {
        return service.provideLocalServerService()
            .flatMapConcat {
                it.fsGet(dir)
                    .map(extractResponseWithErrorHandling())
            }
    }

    override fun uploadAudio(hash: String?): Flow<Int> {
        return service.provideLocalServerService()
            .flatMapConcat {
                it.uploadAudio(
                    hash,
                    Settings.get().accounts().currentAccessToken,
                    UserAgentTool.USER_AGENT_CURRENT_ACCOUNT
                )
                    .map(extractResponseWithErrorHandling())
            }
    }

    companion object {
        inline fun <reified T : Any> extractResponseWithErrorHandling(): (BaseResponse<T>) -> T =
            { r ->
                r.error?.let {
                    throw Exception(
                        firstNonEmptyString(
                            r.error?.errorMsg,
                            "Error"
                        )
                    )
                }
                r.response ?: throw NullPointerException("Local Server return null response")
            }
    }
}